### ISSUE TITLE
GrpcStreamingCall.executeIn() channels are not closed on errors

### ISSUE THREAD

**Issue by alexvasilkov** [square/wire#1953]
I have a streaming gRPC method declared like this:

```
rpc Sync(SyncRequest) returns (stream SyncBlock) {}
```

And I'm trying to use generated gRPC client like this:

```kotlin
  suspend fun sync() = coroutineScope {
    try {
      val syncState = readSyncState()

      val (sender, receiver) = grpc.Sync().executeIn(this)

      sender.send(SyncRequest(syncState))

      for (response in receiver) {
        saveSyncResponse(response)
      }
    } catch (th: Throwable) {
      delay(2000L)
      sync() // Retry until cancelled
    }
  }
```

The idea is that `sync()` method will be called and cancelled by application scope whenever appropriate, while the method itself will suspend forever and should handle reconnects.

I can successfully connect and start receiving sync events, but if connection is interrupted (e.g. enable flight mode) I see an okhttp stractrace in logs but the error is not thrown, `receiver` is not closed and `sync()` is hanging forever in the `for` loop.

It turns out that `GrpcStreamingCall.executeIn()` does not handle IO exceptions properly (see [readFromResponseBodyCallback](https://github.com/square/wire/blob/682f863/wire-library/wire-grpc-client/src/jvmMain/kotlin/com/squareup/wire/internal/grpc.kt#L94)) and once an exception is thrown from `onResponse(...)` method it will be ignored by okhttp (see [RealCall](https://github.com/square/okhttp/blob/4677bee/okhttp/src/main/kotlin/okhttp3/internal/connection/RealCall.kt#L523)) and neither `receiver` nor `sender` channels won't be closed.

In contrast `GrpcStreamingCall.executeBlocking()` correctly handles the errors in `onResponse(...)` (see [BlockingMessageSource](https://github.com/square/wire/blob/682f863/wire-library/wire-grpc-client/src/jvmMain/kotlin/com/squareup/wire/internal/BlockingMessageSource.kt#L81)).

**Comment by alexvasilkov** [comment]
I ended up using `executeBlocking` but it is not pretty and blocks extra thread while waiting for new messages:

```kotlin
  suspend fun sync(): Unit = coroutineScope {
    try {
      val syncState = readSyncState()

      val call = grpc.Sync()
      val (sink, source) = call.executeBlocking()

      // Wait for coroutine cancellation to cancel the call
      launch { awaitCancellation() }.invokeOnCompletion { call.cancel() }

      withContext(Dispatchers.IO) {
        sink.write(SyncRequest(syncState))
        while (true) {
          val response = source.read() ?: throw IOException("Unexpected stream end")
          saveSyncResponse(response)
        }
      }
    } catch (th: Throwable) {
      delay(2000L)
      sync() // Retry until cancelled
    }
  }
```

**Comment by swankjesse** [comment]
Yikes. Yep, this needs a fix. Thanks for the great bug report.

**Comment by oldergod** [comment]
I think this has been fixed by #2038
It's part of the latest alpha.

https://github.com/square/wire/blob/master/wire-library/wire-grpc-client/src/jvmMain/kotlin/com/squareup/wire/internal/grpc.kt#L96-L124
